### PR TITLE
Memory Usage Fix: XML to JSON transformation from filter to source

### PR DESCRIPTION
**Pull request by vrdmr**
Implemented the logic of transforming XML to JSON from fluentd
filter to source. The change_tracking_runner internally uses the same
library, with some minor changes in the code of library to make it
modular.

@robbiezhang @agup006 @weijiel @Microsoft/omsagent-devs Please review my changes. Thanks.

### CODE REVIEW

File: source/code/plugins/changetracking_lib.rb [hunk 1]
```diff
@@ -130,39 +122,46 @@ def self.transform_and_wrap(inventoryXMLstr, host, time, force_send_run_interval
...
-                    "Timestamp" => timestamp,
-                    "Computer" => host,
-                    "ConfigChangeType"=> "Files",
-                    "Collections"=> fileInventories
+        return {"packages" => packages, "services" => services , "fileInventories" => fileInventories}
+    end
+
+    def self.wrap (time, host, inventory_hash)
+        @@log.debug "The keys in inventory_hash - #{inventory_hash.keys}"
+        @@log.debug "The keys in inventory_hash - #{inventory_hash.keys}"
```
Comment (robbiezhang) [inline source/code/plugins/changetracking_lib.rb]: dup log

File: source/code/plugins/changetracking_lib.rb [hunk 2]
```diff
@@ -130,39 +122,46 @@ def self.transform_and_wrap(inventoryXMLstr, host, time, force_send_run_interval
...
+        wrapper = {}
+        # inventory_hash.key?("inventory_hash")
```
Comment (robbiezhang) [inline source/code/plugins/changetracking_lib.rb]: remove

File: source/code/plugins/changetracking_lib.rb [hunk 3]
```diff
@@ -130,39 +122,46 @@ def self.transform_and_wrap(inventoryXMLstr, host, time, force_send_run_interval
...
+                timestamp = OMS::Common.format_time(time)
+                wrapper = {
+                    "DataType"=>"CONFIG_CHANGE_BLOB",
+                    "IPName"=>"changetracking",
+                    "DataItems"=>[
+                        {
+                            "Timestamp" => timestamp,
+                            "Computer" => host,
+                            "ConfigChangeType"=> "Software.Packages",
+                            "Collections"=> inventory_hash["packages"]
```
Comment (robbiezhang) [inline source/code/plugins/changetracking_lib.rb]: If the key doesn't exist or the value is nil, shall we include it in the DataItems?

File: source/code/plugins/changetracking_lib.rb [hunk 4]
```diff
@@ -130,39 +122,46 @@ def self.transform_and_wrap(inventoryXMLstr, host, time, force_send_run_interval
...
-                    "Collections"=> fileInventories
+        return {"packages" => packages, "services" => services , "fileInventories" => fileInventories}
+    end
+
+    def self.wrap (time, host, inventory_hash)
+        @@log.debug "The keys in inventory_hash - #{inventory_hash.keys}"
+        @@log.debug "The keys in inventory_hash - #{inventory_hash.keys}"
+        wrapper = {}
+        # inventory_hash.key?("inventory_hash")
+        if ((inventory_hash.has_key?("packages") and inventory_hash["packages"].size > 0) or
```
Comment (robbiezhang) [inline source/code/plugins/changetracking_lib.rb]: inventory_hash['packages'] will be nil, if we set inventory_hash['packages'] = nil. Thus, the .size will throw NilClass exception.

File: source/code/plugins/filter_changetracking.rb [hunk 5]
```diff
@@ -33,10 +32,8 @@ def shutdown
     end
 
     def filter(tag, time, record)
-      xml_string = record['xml']
-      @log.debug "ChangeTracking : Filtering xml size=#{xml_string.size}"
-      return ChangeTracking.transform_and_wrap(xml_string, @hostname, time, @force_send_run_interval)
+      @log.debug "ChangeTracking: Got new data to send"
+      return record
```
Comment (robbiezhang) [inline source/code/plugins/filter_changetracking.rb]: why do you need this filter?

Comment (robbiezhang) [inline source/code/plugins/filter_changetracking.rb]: A second thought: this filter should be used as a logger for the change_tracking_runner.rb and changetracking_lib.rb. The current implementation has an issue that the log from these 2 files are not written into the omsagent.log. Can we include the log into the output of the runner as part of the json, and write them into the @log here?

**Comment by gauravga** [comment]
Will you also need to modify tests?

File: source/code/plugins/change_tracking_runner.rb [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+require 'logger'
+require 'digest'
+require 'json'
+
+require_relative 'changetracking_lib'
+require_relative 'oms_common'
+
+
+class ChangeTrackingRunner 
+	CHANGE_TRACKING_FILE = "/etc/opt/omi/conf/omsconfig/configuration/ChangeTrackingInventory.xml"
```
Comment (vrdmr) [inline source/code/plugins/change_tracking_runner.rb]: I'll test with the newer path - /var/opt/Microsoft/omsagent/tmp/changetrackinginventory.xml

File: source/code/plugins/change_tracking_runner.rb [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+			# Read the inventory XML.
+			file = File.open(CHANGE_TRACKING_FILE, "rb")
+			xml_string = file.read; nil # To top the output to show up on STDOUT.
+
+			#Transform the XML to HashMap
+			transformed_hash_map = ChangeTracking.transform(xml_string, @hostname, time, @log)
+			@log.debug("transformed_hash_map size - #{transformed_hash_map["packages"].size}, #{transformed_hash_map["services"].size}, #{transformed_hash_map["fileInventories"].size}")
+
+			output = ChangeTracking.wrap(time, @hostname, transformed_hash_map)
+			@log.debug("output size - #{output.to_s}")
```
Comment (vrdmr) [inline source/code/plugins/change_tracking_runner.rb]: I'l convert the`@log.debug("output size - #{output.to_s}")` to only look for lengths of dataitems.

File: source/code/plugins/filter_changetracking.rb (hunk 5 above)
Comment (vrdmr) [inline source/code/plugins/filter_changetracking.rb]: I'll try to see how it can be done (using the STDERR).

**Comment by vrdmr** [comment]
@gauravga - Yes, I'll make the changes in the unittests.

**Comment by gauravga** [comment]
@vrdmr - I started a new pull request with these changes and mine, merged together.

**Pull request closed without merging**